controllers/node-delete/controller.go

```diff
@@ -0,0 +1,91 @@
+
+import (
+	"github.com/go-logr/logr"
+	"github.com/storageos/api-manager/internal/pkg/storageos"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	// MaxConcurrentReconciles is the maximum number of concurrent Reconciles which can be run.
+	MaxConcurrentReconciles = 5
+)
+
+// Reconciler reconciles a Node object by deleting the StorageOS node object
+// when the corresponding Kubernetes node is deleted.
+type Reconciler struct {
+	client.Client
+	Log logr.Logger
+	api storageos.NodeDeleter
+}
+
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+
+// NewReconciler returns a new Node delete reconciler.
+func NewReconciler(api storageos.NodeDeleter, k8s client.Client) *Reconciler {
+	return &Reconciler{
+		Client: k8s,
+		Log:    ctrl.Log.WithName("controllers").WithName("NodeDelete"),
+		api:    api,
+	}
+}
+
+// SetupWithManager registers the controller with the controller manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{MaxConcurrentReconciles: MaxConcurrentReconciles}).
+		For(&corev1.Node{}).
+		WithEventFilter(ChangePredicate{}).
+		Complete(r)
+}
+
+// Reconcile deletes the StorageOS node.  The delete will fail if the control
+// plane has not yet determined that the node is offline.  Any errors will
+// result in a requeue, with standard back-off retries.
+//
+// Events are not sent as they require an object. By this point the namespace
+// object will no longer be available.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	err := r.api.DeleteNode(req.Name)
+	if err != nil && err != storageos.ErrNamespaceNotFound {
+		// Re-queue without error.  We will get frequent transient errors, such
+		// as version conflicts or locked objects - that's ok.  Even refusal to
+		// delete due to the node still showing as online should be seen as
+		// transient and should not raise an error.
+		return ctrl.Result{Requeue: true}, nil
+	}
+	// We can't return an event as it requires an object, and we can't rely on
+	// the object still being present.
+	return ctrl.Result{}, nil
+}
+
+// ChangePredicate filters events before enqueuing the keys.
+type ChangePredicate struct {
+	predicate.Funcs
+}
+
+// Create determines whether an object create should trigger a reconcile.
+func (ChangePredicate) Create(event.CreateEvent) bool {
+	return false
+}
+
+// Update determines whether an object update should trigger a reconcile.
+func (ChangePredicate) Update(event.UpdateEvent) bool {
+	return false
+}
+
+// Delete determines whether an object delete should trigger a reconcile.
+func (ChangePredicate) Delete(e event.DeleteEvent) bool {
+	// Ignore objects without metadata.
+	return e.Meta != nil
```

Comment:
We should also be able to check the node annotations to find out if storageos is installed on the node and ignore the events from nodes that don't have storageos.